### PR TITLE
add allocated disk size metric

### PR DESCRIPTION
**Pull request by avlitman**
### What this PR does
add allocated disk size metric in order to monitor vms sizes.

Before this PR:
kubevirt_vm_disk_allocated_size_bytes does not exists
After this PR:
kubevirt_vm_disk_allocated_size_bytes exists

Fixes # 
jira-ticket: https://issues.redhat.com/browse/CNV-49002

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer
Metric output:

Screenshot from 2024-11-04 14-15-08

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
kubevirt_vm_disk_allocated_size_bytes metric added in order to monitor vm sizes
```

### CODE REVIEW

File: pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go
Comment (mhenriks) [inline pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go:63]: should check `volume.DataVolume != nil` as well

File: pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go
Comment (mhenriks) [inline pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go:75]: Why do you have to get the PV?

File: pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go
Comment (mhenriks) [inline pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go:81]: Using capacity is fine for block PVC but does not work for filesystem PVCs on shared storage (kubevirt hpp or local storage for example). For that case you have to get the size of the `disk.img` file on the PVC

File: docs/metrics.md
Comment (mhenriks) [inline docs/metrics.md:90]: How do you plan to get this data?  Do you need guest agent integration?

**Review by mhenriks — COMMENTED** [review]
Couple edge cases to consider and very curious how you plan to get used bytes

File: pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go:75]: since I need the pvName for the metric

File: docs/metrics.md
Comment (avlitman) [reply docs/metrics.md:90]: This query returns the used size by volume name, namespace, but its missing the connection to the vm name.

sum by (name, namespace)(kubelet_volume_stats_used_bytes + on (persistentvolumeclaim,namespace) group_left(name)(label_replace(0*kube_persistentvolumeclaim_info , "persistentvolume",  "$1",  "volumename",  "(.*)") +on (persistentvolume) group_left(name)(0*kube_persistentvolume_claim_ref)))


so to be honest I'm not sure yet if I should use this metric and somehow add the vm name or get the used size directly. any ideas?

File: pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go:81]: I thought that this metric covers the filesystem bytes https://github.com/kubevirt/kubevirt/blob/main/docs/metrics.md#kubevirt_vmi_filesystem_capacity_bytes, no?

Comment (mhenriks) [reply pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go:81]: Those stats are from guest perspective. Also it is an aggregate.

File: docs/metrics.md
Comment (mhenriks) [reply docs/metrics.md:90]: I can't imagine that query works for block PVCs.

I don't think there is any standard way you can get used bytes for a block device from k8s. You could get free space from guest perspective but if user "preallocated" the DataVolume than actual used will likely be greater

For filesystem PVCs, you can stat the disk.img file to get accurate used/free byte stats that will reflect preallocation

Comment (avlitman) [reply docs/metrics.md:90]: ohhh so that's a problem.. thanks for the info.

@sradco WDYT?

File: pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go:81]: +1 so I will add also the filesystem PVCs thanks a lot!

File: docs/metrics.md
Comment (machadovilaca) [reply docs/metrics.md:90]: something like https://libvirt.org/manpages/virsh.html#domfsinfo would work?

File: pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmdisk_metrics.go:81]: @mhenriks sorry for bothering again, do you have maybe a reference how to get the disk.img file size? struggling to understand

File: docs/metrics.md
Comment (avlitman) [reply docs/metrics.md:90]: @mhenriks WDYT about joao comment?

File: docs/metrics.md
Comment (sradco) [inline docs/metrics.md:87]: @avlitman  @machadovilaca  I think it might make sense to add this to the kubevirt_vm_resource_requests metric. Wdyt?

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (sradco) [inline pkg/monitoring/metrics/virt-controller/vmstats_collector.go:43]: @avlitman please check the how the pvc name label is called also in other metrics, to make sure we are aligned.

File: docs/metrics.md
Comment (machadovilaca) [reply docs/metrics.md:87]: I like, and few changes would be needed from the current impl

File: pkg/monitoring/metrics/virt-controller/metrics.go
Comment (machadovilaca) [inline pkg/monitoring/metrics/virt-controller/metrics.go:59]: you need to add the new informer args in `pkg/virt-controller/watch/application.go:435`

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (machadovilaca) [inline pkg/monitoring/metrics/virt-controller/vmstats_collector.go:44]: maybe he should try to keep the order in `Metrics:         append(timestampMetrics, vmResourceRequests, vmInfo, vmPVCAllocatedSize),` going forward

please create this metric last

File: pkg/monitoring/metrics/virt-controller/metrics.go
Comment (machadovilaca) [inline pkg/monitoring/metrics/virt-controller/metrics.go:41]: why remove this?

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (machadovilaca) [inline pkg/monitoring/metrics/virt-controller/vmstats_collector.go:428]: this function should be refactored imo, it's too big
maybe try to move the body of the loops into separate function

File: pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go
Comment (machadovilaca) [inline pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go:42]: unit tests are failing because you are probably incorrectly configuring the informer for the collector, check the implementation in https://github.com/kubevirt/kubevirt/blob/d7420c6e2eb3a3029847b36ffaa48bb0cdf8b00e/pkg/monitoring/metrics/virt-controller/vmistats_collector_test.go#L268

**Review by machadovilaca — CHANGES_REQUESTED** [review]

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:71]: thanks!

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go]: so I update all metric to be according to the order

File: pkg/monitoring/metrics/virt-controller/metrics.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/metrics.go:41]: recent rebase sorry, fixed

File: pkg/monitoring/metrics/virt-controller/metrics.go
Comment (mhenriks) [inline pkg/monitoring/metrics/virt-controller/metrics.go:153]: Why would you need this? Informers already cache the resource by namespace/name. Should be able to do something like:
```golang
persistentVolumeClaimInformer.GetStore().GetByKey("ns/name")
```

**Review by mhenriks — COMMENTED** [review]
I suggest getting rid of the index on PVC informer. You shouldn't need it.

For future reference if you do need indexes, you have to use `testutils.NewFakeInformerWithIndexersFor` so the fake informer is aware of the indexes

File: pkg/monitoring/metrics/virt-controller/metrics.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/metrics.go]: thanks a lot finally able to get it working! now adding more labels

**Comment by avlitman** [comment]
/hold

e2e tests still not working as expected

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (mhenriks) [inline pkg/monitoring/metrics/virt-controller/vmstats_collector.go:538]: Why going through this list to find a single PVC when you can call `.GetStore().GetByKey("namespace/name")`

Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:538]: I'm doing it in this line:https://github.com/kubevirt/kubevirt/pull/12765/files#diff-1f9be9d6f2363d168e35db520321088d1d544b5c7afd205f177d002f46f13129R506

not sure why here no, I will fix

Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go]: fixed to use one func that uses `.GetStore().GetByKey("namespace/name")`. @mhenriks will appreciate you lgtm if it looks good

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (mhenriks) [inline pkg/monitoring/metrics/virt-controller/vmstats_collector.go:513]: Can you be more specific about what you mean by "allocated disk size"? I'm not sure this is the value you are looking for. This is the size of PVC that was requested. It does not map directly to the size of disk image used by the VM

Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:513]: @mhenriks allocated size means the maximum size the disk can use

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (sradco) [inline pkg/monitoring/metrics/virt-controller/vmstats_collector.go:162]: @avlitman
```suggestion
			Help: "Allocated disk size of a Virtual Machine in bytes, based on its PersistentVolumeClaim. Includes persistentvolumeclaim (PVC name), volume_mode (disk presentation mode: Filesystem or Block), and device (disk name).",
```

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (sradco) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:513]: Just a small correction. That is not the main use case.
Knowing the allocated size is important for many use cases like:
1. Resources capacity planning - See how much is allocated and plan for next year for example.
2. Resources utilization - We can calculate the used size out of the total
3. Resources management -  Free up unused storage

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go]: done

File: tests/monitoring/metrics.go
Comment (fossedihelm) [inline tests/monitoring/metrics.go:26]: please move this import in the proper section below

File: pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go
Comment (fossedihelm) [inline pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go:379]: I think unit tests should cover several scenarios: empty volumeMode is just an example

**Review by fossedihelm — COMMENTED** [review]
Thank you @avlitman 
Some comments below

File: pkg/monitoring/metrics/virt-controller/metrics.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/metrics.go:51]: it's not in this pr scope and require many manual verification. we can open an issue/ticket for that

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:166]: Yes since it's also in kubernetes metrics, this is the used term, asked by @sradco.

It was disk_name at the begging of this pr..

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:506]: fixed thanks (:

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:509]: fixed

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:513]: fixed

File: tests/monitoring/metrics.go
Comment (avlitman) [reply tests/monitoring/metrics.go:26]: fixed

File: pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go]: added more tests

**Comment by avlitman** [comment]
HI @fossedihelm thanks a lot! I fixed all your comments will appreciate if you can review again and hopefully approve/lgtm (:

**Comment by avlitman** [comment]
HI @fossedihelm thanks a lot! I fixed all your comments will appreciate if you can review again and hopefully approve/lgtm (:

File: pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go
Comment (fossedihelm) [inline pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go:399]:
```suggestion
            pvc := createPVC("default", "test-vm-pvc", resource.MustParse("5Gi"), pointer.P(k8sv1.PersistentVolumeFilesystem))
```

File: pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go
Comment (fossedihelm) [inline pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go:476]: same as above

File: pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go
Comment (fossedihelm) [inline pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go:474]: I don't think this differ from `should collect PVC size metrics correctly`. Can be dropped

**Review by fossedihelm — COMMENTED** [review]
Just some very small things and we are ready to go from my point of view

File: pkg/monitoring/metrics/virt-controller/metrics.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/metrics.go:51]: opened: https://github.com/kubevirt/kubevirt/issues/13173 can you please review?

File: pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go:399]: fixed

File: pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector_test.go:476]: func removed

**Comment by avlitman** [comment]
> Just some very small things and we are ready to go from my point of view

@fossedihelm thank you so much! fixed all (:

**Comment by fossedihelm** [comment]
/lgtm
Thanks

File: pkg/monitoring/metrics/virt-controller/metrics.go
Comment (fossedihelm) [reply pkg/monitoring/metrics/virt-controller/metrics.go:51]: looks good, thanks

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (machadovilaca) [inline pkg/monitoring/metrics/virt-controller/vmstats_collector.go:164]: nit: please consistently add space to either the end of beginning of lines

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (machadovilaca) [inline pkg/monitoring/metrics/virt-controller/vmstats_collector.go:492]: nit:
```suggestion
		if vm.Spec.Template != nil {
			cr = append(cr, collectDiskMetricsFromPVC(vm)...)
		}
```

**Review by machadovilaca — APPROVED** [review]
just a couple nits, feel free to address them later

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:492]: fixed (: thanks!

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:164]: fixed

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (fossedihelm) [inline pkg/monitoring/metrics/virt-controller/vmstats_collector.go:187]: Consider making these unexported

Comment (machadovilaca) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:187]: at the time I started exporting these functions so that we could unlink the unit test code from the product code

Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go]: absolutely right! fixed

Comment (fossedihelm) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go]: It just depends on if we want to perform black-box or white-box testsing.

Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go]: I think that if we decide to move the tests then we can change it again to be exported. is it ok @machadovilaca?

Comment (avlitman) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:187]: ok so @machadovilaca mentioned unit test will fail if changing it, so I fixed it back to be exported.
we can discusses it here and if we decide to change we can do it in a separate pr

**Comment by machadovilaca** [comment]
> ok so @machadovilaca mentioned unit test will fail if changing it, so I fixed it back to be exported. we can discusses it here and if we decide to change we can do it in a separate pr

/lgtm

File: pkg/monitoring/metrics/virt-controller/vmstats_collector.go
Comment (fossedihelm) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:187]: Ok but unit tests don't fail, if proper renamed:
```
Will run 72 of 72 specs
••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••;5;10m••••••••••••

Ran 72 of 72 Specs in 0.002 seconds
SUCCESS! -- 72 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestVirtController (0.00s)
PASS

Process finished with the exit code 0
```

Comment (machadovilaca) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:187]: > It just depends on if we want to perform black-box or white-box testing.

what would be the disadvantage of this approach? these are called directly in the Prometheus callback, we just skip the `vmInformer.GetIndexer().List()` in the tests basically
and this way we can test each specific collect individually

what would be your suggestion? unexporting these functions and test the callback as a whole?

Comment (fossedihelm) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:187]: To me, we can just perform white-box testing, which means unexporting the functions and test'em as we are doing right now in the test file.
We are already doing it in most of the controller tests.
The issue with unexported functions will appear only if you want to switch the test package to `*_test` (black-box testing). But it's up to you, I will not force anyone to use white-box testing. There are advantages and disadvantages to both strategies.

Comment (machadovilaca) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:187]: yes, I was looking into switching to `*_test`, I thought it was the direction we were trying to go, saw some PRs with those kind of changes

Comment (fossedihelm) [reply pkg/monitoring/metrics/virt-controller/vmstats_collector.go:187]: I think we should open a wider discussion to define what is our direction :)

**Pull request merged**
